[SimplifyCFG] Extend jump-threading to allow live local defs

LU-JOHN opened the pull request:
Extend jump-threading to allow local defs that are live outside of the threaded block.  Allow threading to destinations where the local defs are not live.

LU-JOHN committed:
Extend jump-threading to allow live local defs

Signed-off-by: John Lu <John.Lu@amd.com>

LU-JOHN committed:
Avoid two lookups

Signed-off-by: John Lu <John.Lu@amd.com>

dtcxzyw reviewed (commented):
It is not obvious to judge whether this patch is profitable. Can you share some performance data on SPEC/LLVM test-suite/your motivating case?

LU-JOHN committed:
Use using instead of typedef

Signed-off-by: John Lu <John.Lu@amd.com>

LU-JOHN committed:
Do expensive reachesUsed test after inexpensive checks

Signed-off-by: John Lu <John.Lu@amd.com>

LU-JOHN commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3467,13 +3485,23 @@ static bool blockIsSimpleEnoughToThreadThrough(BasicBlock *BB) {
 
+  for (BasicBlock *Succ : successors(BB)) {
+    BlocksSet VisitedBlocksSet;
+    if (reachesUsed(Succ, UsedInNonLocalBlocksSet, VisitedBlocksSet))
```
Do reachesUsed after check checks.

LU-JOHN committed:
If we determine a block reaches non-local uses, record that result.

Signed-off-by: John Lu <John.Lu@amd.com>

LU-JOHN committed:
Search for live blocks backward from uses.  Add limit for searching

Signed-off-by: John Lu <John.Lu@amd.com>

LU-JOHN commented:
> It is not obvious to judge whether this patch is profitable. Can you share some performance data on SPEC/LLVM test-suite/your motivating case?

This PR was inspired by code generated for Triton.  A reduced example can be found in function testA in jump-threading-live-on-exit.ll.

I've changed the block searching to go backward from uses to the def instead of forward from the def.   This is more efficient since a forward search is not guaranteed to hit a use, but a backward search must hit the def.  I've also added a limit to ensure that the search gives up if too many blocks are searched.

I tested this with a 3-stage build of LLVM.  This change increased the number of threaded branches from` 186726 to 197086, +10360.  The number of blocks that could be threaded and required backward searches from the uses was 331442. The number of blocks searched from the uses ranged from 0 to 25 (the search limit).  The average of number of blocks searched was 5.21.

LU-JOHN committed:
Update test results

Signed-off-by: John Lu <John.Lu@amd.com>

dtcxzyw commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -3540,18 +3568,29 @@ foldCondBranchOnValueKnownInPredecessorImpl(BranchInst *BI, DomTreeUpdater *DTU,
+  if (!blockIsSimpleEnoughToThreadThrough(BB, NonLocalUseBlocks))
+    return false;
+
+  if (NonLocalUseBlocks.contains(BI->getSuccessor(0)) &&
```
Is this a heuristic to check if the non-local uses live in both successors?
Can you add more comments for this check?

LU-JOHN replied:
Comments added to describe a quick check before doing more expensive search.

dtcxzyw reviewed (commented):
@nikic Can you check the compile-time impact of this patch?

LU-JOHN committed:
Add more comments

Signed-off-by: John Lu <John.Lu@amd.com>

dtcxzyw reviewed (approved):
LGTM.
